[spirv] add predefined macro '__spirv'

jaebaek opened the pull request.

sudonatalie commented:
Would it be too much trouble to add a DXIL codegen test that makes sure these lines are removed when DXC is run without `-spirv`?

jaebaek commented:
Good point. Added.

@pow2clk @tex3d This is a heads-up. I added the implicit definition of `__spirv` and a DXIL test for it. Please let me know if the file location for the DXIL test I used is wrong.

jaebaek commented on docs/SPIR-V.rst:
```diff
@@ -358,6 +358,21 @@ interface variables:
+
+.. code:: hlsl
+
+  #ifdef __spirv
```
TODOs based on offline discussion:
* Move it to https://github.com/microsoft/DirectXShaderCompiler/tree/master/tools/clang/test/HLSLFileCheck/hlsl/preprocessor/version_defines 
* Change `__spirv` to `__SPIRV`
